src/Grumplestiltskin/EllipticCurve.hs

```diff
@@ -0,0 +1,175 @@
+If you benchmark the execution time, you will see this version runs much quicker,
+and it is the implementation used by pyEvm. For educational purposes however,
+it is preferable to use the non-optimized version as it structures the points in a more intuitive way (the usual x, y tuple).
+The optimized version structures EC points as 3-tuples, which are harder to interpret.
+
+13.2 section in Handbook of Elliptic and Hyperelliptic Curve Cryptography -- Henri Cohen
+-}
+
+newtype PECPoint (s :: S) = PECPoint (Term s (PPair PGFElement PGFElement))
+    deriving stock
+        (Generic)
+    deriving anyclass
+        (SOP.Generic)
+    deriving
+        (PlutusType)
+        via (DeriveNewtypePlutusType PECPoint)
+
+instance PEq PECPoint
+
+newtype PECIntermediatePoint (s :: S) = PECIntermediatePoint (Term s (PPair PGFIntermediate PGFIntermediate))
+    deriving stock
+        (Generic)
+    deriving anyclass
+        (SOP.Generic)
+    deriving
+        (PlutusType)
+        via (DeriveNewtypePlutusType PECIntermediatePoint)
+
+pToPoint :: forall (s :: S). Term s PPositive -> Term s PECIntermediatePoint -> Term s PECPoint
+pToPoint fieldModulus p =
+    pcon $
+        PECPoint $
+            pmatch
+                (pupcast p)
+                (\(PPair x y) -> pcon $ PPair (pgfToElem x fieldModulus) (pgfToElem y fieldModulus))
+
+{- | @P + Q = R@ where @R@ is the inverse of a point on the intersection of the curve and the line defined by @P@ and @Q@.
+In case @P == Q@, we calculate the @2P@.
+-}
+pAddPoints :: forall (s :: S). Term s PPositive -> Term s PInteger -> Term s PECIntermediatePoint -> Term s PECIntermediatePoint -> Term s PECIntermediatePoint
+pAddPoints fieldModulus curveA point1 point2 =
+    pmatch
+        (pupcast point1)
+        ( \(PPair pointX1 pointY1) ->
+            pmatch
+                (pupcast point2)
+                ( \(PPair pointX2 pointY2) ->
+                    let
+                        -- TODO: maybe for the `yDiff` and `xDiff`, the `plet` should be used?
+                        yDiff = pointY1 #- pointY2
+                        xDiff = (pointX1 #- pointX2)
+                     in
+                        pif
+                            -- Are the points equal?
+                            -- Doing the modulo normalisation 2x as part of `pgfToElem`. Can we do better?
+                            -- TODO: handle the case when X coordinates of both points are the same.
+                            (pgfToElem yDiff fieldModulus #== pgfZero #&& pgfToElem xDiff fieldModulus #== pgfZero)
+                            (pPointDouble fieldModulus curveA point1)
+                            ( let lambdaNum = yDiff
+                                  lambdaDenum = pgfFromElem $ pgfRecip # xDiff # fieldModulus
+                                  lambda = lambdaNum #* lambdaDenum
+                                  pointX3 = ((lambda #* lambda) #- pointX1) #- pointX2
```

Comment:
`lambda #* lambda` is `lambda` squared, which might be faster to do as an exponentiation op, as this calls out to a builtin.